src/artifact/parser.cpp

```diff
@@ -56,25 +56,25 @@ ExpectedArtifact VerifyEmptyPayloadArtifact(
 	}
 	// TODO - When augmented sections are added - check for these also
 	log::Trace("Empty payload Artifact: Verifying empty payload");
-	auto tok = lexer.Next();
-	if (tok.type == token::Type::Payload) {
+	auto expected_payload = artifact.Next();
+	if (!expected_payload
+		&& expected_payload.error().code
+			   != parser_error::MakeError(parser_error::EOFError, "").code) {
+		return expected::unexpected(
+			expected_payload.error().WithContext("While verifying empty payload"));
+	} else {
 		// If a payload is present, verify that it is empty
-		auto expected_payload = artifact.Next();
-		if (!expected_payload) {
-			auto err = expected_payload.error();
-			if (err.code.value() != parser_error::Code::NoMorePayloadFilesError) {
-				return expected::unexpected(parser_error::MakeError(
-					parser_error::Code::ParseError,
-					"This should never happen, we have a payload token / programmer error"));
-			}
-			return artifact;
-		}
 		auto payload = expected_payload.value();
```

Comment:
Q: What if we get `EOFError`?

Reply:
Good catch, I'm rearranging the `if` to deal with this.